src/Microsoft.AspNet.OData.Shared/TypeHelper.cs

```diff
@@ -491,5 +491,75 @@ private static Type GetInnerGenericType(Type interfaceType)
 
             return null;
         }
+
+        /// <summary>
+        /// Searches the public property with the specified name from the target type and its base types safely,
+        /// circumventing <see cref="AmbiguousMatchException"/> where a property has been redeclared in a derived type.
+        /// </summary>
+        /// <param name="targetType">The type to search from.</param>
+        /// <param name="propertyName">The name of the public property to get.</param>
+        /// <returns>
+        /// An object representing the public property with the specified name from the most derived class,
+        /// if found; otherwise, null.
+        /// </returns>
+        internal static PropertyInfo GetProperty(Type targetType, string propertyName)
+        {
+            // NOTE: Designed to avoid costly closure lambda allocations since method might be used in hot path
+            List<PropertyInfo> matchedProperties = null;
+
+            PropertyInfo[] properties = targetType.GetProperties();
+            for (int i = 0; i < properties.Length; i++)
+            {
+                PropertyInfo propertyInfo = properties[i];
+
+                if (!propertyInfo.Name.Equals(propertyName))
+                {
+                    continue;
+                }
+
+                // Property declared on target type - opportunity to exit early
+                if (propertyInfo.DeclaringType.Equals(targetType))
+                {
+                    return propertyInfo;
+                }
+
+                // We defer initialization of the list since we (arguably) don't get this far
+                if (matchedProperties == null)
+                {
+                    matchedProperties = new List<PropertyInfo>(4);
```

Comment:
nitpick: This is probably overkill. I think in the majority of the cases, I think only one property will be returned. In that case maybe we can avoid creating the list and store it in a variable and only create the list when the second property is found. But I don't know if that is worth the code complexity. In any case, not a blocker.

Reply:
Fixed. See my comment [here](https://github.com/OData/WebApi/pull/2523#discussion_r809903958)